### PR TITLE
Replaced the hardcoded certificates used in crypto_crls_cert_chains test with dynamically generated certificates

### PR DESCRIPTION
**Pull request by soccerGB**
The changes in this PR addresses the following issue.

https://github.com/Microsoft/openenclave/issues/783

### CODE REVIEW

**Comment by soccerGB** [comment]
bors try

File: tests/crypto_crls_cert_chains/host/CMakeLists.txt [hunk 1]
```diff
@@ -3,12 +3,85 @@
...
+            # The certificates and CRLs were generated using the following openssl commands.
+            # Creating Root CA private key and self-signed Root CA cert
+            COMMAND openssl genrsa -out ${DATA_DIR}/RootCA1.key.pem
+            COMMAND openssl req -new -x509 -key ${DATA_DIR}/RootCA1.key.pem -out ${DATA_DIR}/RootCA1.crt.pem -days 3650 -subj "/C=US/ST=Ohio/L=Columbus/O=Acme Company/OU=Acme/CN=Root"
+
+            # We want to generate certificates used in this tests with at least one second apart to make it possible to sort those certificate by time 
+            # "Not Before" of a x509 certificate, which has time resolution up to second 
+            # Creating Intermediate CA certificate, signed by the Intermediate CA
+            COMMAND sleep 1
+            COMMAND openssl genrsa -out ${DATA_DIR}/IntermediateCA1.key.pem
```
Comment (anakrish) [inline tests/crypto_crls_cert_chains/host/CMakeLists.txt]: @andschwa Is it better to use many cmake COMMANDs or is it better to have a bash script and then call it to do stuff?

**Comment by soccerGB** [comment]
The current certificate "does" handle   different order of certificates but not the 1 second requirement, which will be addressed later

File: tests/crypto_crls_cert_chains/host/CMakeLists.txt [hunk 2]
```diff
@@ -3,12 +3,85 @@
...
+            COMMAND openssl crl -inform pem -outform der -in ${DATA_DIR}/intermediate_crl1.pem -out ${DATA_DIR}/intermediate_crl1.der
+            COMMAND openssl crl -inform pem -outform der -in ${DATA_DIR}/intermediate_crl2.pem -out ${DATA_DIR}/intermediate_crl2.der
+            )
 
 target_compile_options(crypto-extra_host PRIVATE
     -std=c++11
 )
 
-target_link_libraries(crypto-extra_host oehostapp)
+target_link_libraries(crypto-extra_host oehostapp)
```
Comment (anakrish) [inline tests/crypto_crls_cert_chains/host/CMakeLists.txt]: newline needed.

File: tests/crypto_crls_cert_chains/data/intermediate2.cnf [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+[ ca ]
+default_ca = CA_default        # The default ca section
+
+####################################################################
+[ CA_default ]
+database    = ./intermediate_index2.txt
+crlnumber   = ./intermediate_crl_number2
+
+# The root key and root certificate.
+private_key       = ../data/IntermediateCA1.key.pem
```
Comment (anakrish) [inline tests/crypto_crls_cert_chains/data/intermediate2.cnf]: A bit confusing:
intermediate.cnf and intermediate2.cnf uses the same intermediateca1.key and cert.

File: tests/crypto_crls_cert_chains/data/root2.cnf [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+[ ca ]
+default_ca    = CA_default        # The default ca section
+
+####################################################################
+[ CA_default ]
+database    = ./root_index2.txt
+crlnumber   = ./root_crl_number2  # For certificate revocation lists
+
+# The root key and root certificate.
+private_key       = ../data/RootCA1.key.pem
```
Comment (anakrish) [inline tests/crypto_crls_cert_chains/data/root2.cnf]: Is this intended to use the same key and crt?

File: tests/crypto_crls_cert_chains/host/CMakeLists.txt (hunk 2 above)
Comment (soccerGB) [reply tests/crypto_crls_cert_chains/host/CMakeLists.txt]: Addressed

File: tests/crypto_crls_cert_chains/host/CMakeLists.txt [hunk 5]
```diff
@@ -3,12 +3,86 @@
...
+            COMMAND openssl crl -inform pem -outform der -in ${DATA_DIR}/intermediate_crl1.pem -out ${DATA_DIR}/intermediate_crl1.der
+            COMMAND openssl crl -inform pem -outform der -in ${DATA_DIR}/intermediate_crl2.pem -out ${DATA_DIR}/intermediate_crl2.der
+            )
 
 target_compile_options(crypto-extra_host PRIVATE
     -std=c++11
 )
 
 target_link_libraries(crypto-extra_host oehostapp)
+
```
Comment (andschwa) [inline tests/crypto_crls_cert_chains/host/CMakeLists.txt]: Too many newlines 👅

File: tests/crypto_crls_cert_chains/host/CMakeLists.txt (hunk 1 above)
Comment (soccerGB) [reply tests/crypto_crls_cert_chains/host/CMakeLists.txt]: I will keep it as it for now. When it comes to porting this CMake file for Windows, we will revisit this topic again-- whether to wrap them into a script or not.

File: tests/crypto_crls_cert_chains/host/CMakeLists.txt [hunk 6]
```diff
@@ -3,12 +3,86 @@
...
+            COMMAND openssl genrsa -out ${DATA_DIR}/IntermediateCA1.key.pem
+            COMMAND openssl req -new -key ${DATA_DIR}/IntermediateCA1.key.pem -out ${DATA_DIR}/IntermediateCA1.csr -subj "/C=US/ST=Ohio/L=Columbus/O=Acme Company/OU=Acme/CN=Intermediate"
+            COMMAND openssl x509 -req -in ${DATA_DIR}/IntermediateCA1.csr -CA ${DATA_DIR}/RootCA1.crt.pem -CAkey ${DATA_DIR}/RootCA1.key.pem -CAcreateserial -out ${DATA_DIR}/IntermediateCA1.crt.pem -days 3650
+
+            # Creating Leaf certificates signed by the Intermediate CA
+            # Creating Leaf certificates signed by the Intermediate CA
+            COMMAND sleep 1
+            COMMAND echo 'Creating Leaf certificates signed by the Intermediate CA....'
+            COMMAND openssl genrsa -out ${DATA_DIR}/Leaf1.key.pem
+            COMMAND openssl req -new -key ${DATA_DIR}/Leaf1.key.pem -out ${DATA_DIR}/Leaf1.csr -subj "/C=US/ST=Ohio/L=Columbus/O=Acme Company/OU=Acme/CN=Leaf1"
```
Comment (anakrish) [inline tests/crypto_crls_cert_chains/host/CMakeLists.txt]: nice way to automate!

File: tests/crypto_crls_cert_chains/host/CMakeLists.txt [hunk 7]
```diff
@@ -3,12 +3,86 @@
...
+            COMMAND openssl ca -gencrl -config ${DATA_DIR}/root2.cnf -out ${DATA_DIR}/root_crl2.pem> /dev/null
+            COMMAND openssl ca -gencrl -config ${DATA_DIR}/intermediate.cnf -out ${DATA_DIR}/intermediate_crl1.pem
+            COMMAND openssl ca -gencrl -config ${DATA_DIR}/intermediate2.cnf -out ${DATA_DIR}/intermediate_crl2.pem
+
+            # Revoking intermediate and leaf certs
+            COMMAND openssl ca -revoke ${DATA_DIR}/IntermediateCA1.crt.pem -keyfile ${DATA_DIR}/RootCA1.key.pem -cert ${DATA_DIR}/RootCA1.crt.pem -config ${DATA_DIR}/root2.cnf
+            COMMAND openssl ca -revoke ${DATA_DIR}/Leaf2.crt.pem -keyfile ${DATA_DIR}/IntermediateCA1.key.pem -cert ${DATA_DIR}/IntermediateCA1.crt.pem -config ${DATA_DIR}/intermediate2.cnf
+
+            # recreate updated CRL lists
+            COMMAND openssl ca -gencrl -config ${DATA_DIR}/root2.cnf -out ${DATA_DIR}/root_crl2.pem
```
Comment (anakrish) [inline tests/crypto_crls_cert_chains/host/CMakeLists.txt]: The root2.cnf and intermediate2.cnf seem unnecessary.
It seems like you can pass root.cnf and intermediate.cnf here and it should still work.

The crl_number and the index.txt files specified in the cnf act as the database.

Comment (soccerGB) [reply tests/crypto_crls_cert_chains/host/CMakeLists.txt]: I tried that and it involved deleting and recreating  index.txt for each crl creation, which makes it hard to understand CRL creation process. That's why I decided to totally separate the generation of those two CRLs.

**Pull request merged**